### PR TITLE
Add support for Python 3

### LINKED ISSUE #19: Python 3

### PR DESCRIPTION
**Pull request by jni**
Closes #19 

There were two things I wasn't able to fix:

- `TestBilateralFilter.test_01_01_image` is failing for reasons unknown.
- I wasn't able to get `centrosome/src/_cpmorphology.c` to compile as a Python3 extension. I've instead wrapped `scipy.ndimage.maximum_filter` at the appropriate place and removed that file from `setup.py`. I've left it there just in case.

If someone could review this I would much appreciate it!

### CODE REVIEW

File: centrosome/bg_compensate.py [hunk 1]
```diff
@@ -403,7 +406,7 @@ def bg_compensate(img, sigma, splinepoints, scale):
     import time
     t0 = time.clock()
     bg = backgr(img, mask, MODE_AUTO, sigma, splinepoints=splinepoints, scale=scale)
-    print ("Executed in %f sec" % (time.clock() - t0))
+    print(("Executed in %f sec" % (time.clock() - t0)))
```
Comment (AetherUnbound) [inline centrosome/bg_compensate.py]: The double parenthesis here can be removed

File: centrosome/filter.py [hunk 2]
```diff
@@ -1196,7 +1199,7 @@ def line_integration(image, angle, decay, sigma):
     for i in range(1,result_fwd.shape[0]):
         result_fwd[i] += result_fwd[i-1] * decay
     result_rev = smoothed.copy()
-    for i in reversed(range(result_rev.shape[0]-1)):
+    for i in reversed(list(range(result_rev.shape[0]-1))):
```
Comment (AetherUnbound) [inline centrosome/filter.py]: This `list` casting should be redundant

File: centrosome/neighmovetrack.py [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
+from . import lapjv
+import six
+from six.moves import range
+from six.moves import zip
 
 invalid_match = 1000000  # limiting the choices of the algorithms
 
 
-def euclidean_dist((x1, y1), (x2, y2)):
+def euclidean_dist(xxx_todo_changeme, xxx_todo_changeme1):
```
Comment (AetherUnbound) [inline centrosome/neighmovetrack.py]: I assume this still needs to be changed?

File: centrosome/neighmovetrack.py [hunk 4]
```diff
@@ -85,15 +91,15 @@ def from_labels(labels):
...
+        areas = scipy.ndimage.measurements.sum(labels != 0, labels, list(range(1, numpy.max(labels) + 1)))
 
         existing_labels = [i for (i, a) in enumerate(areas, 1) if a > 0]
 
         existing_areas = [a for a in areas if a > 0]
 
         existing_centers = scipy.ndimage.measurements.center_of_mass(labels != 0, labels, existing_labels)
 
-        zipped = (zip(existing_labels, existing_centers, existing_areas))
+        zipped = (list(zip(existing_labels, existing_centers, existing_areas)))
```
Comment (AetherUnbound) [inline centrosome/neighmovetrack.py]: Both the surrounding parenthesis and the `list` cast can be removed from here and above

File: tests/test_cpmorphology.py [hunk 5]
```diff
@@ -1165,7 +1169,7 @@ def test_01_02_circle(self):
         labels = np.zeros((1001,1001),int)
         y,x = np.mgrid[-500:501,-500:501]
         labels[x*x+y*y<=250000] = 1
-        extents = morph.calculate_extents(labels,[1])
+        extents = morph.calculate_extents(labels,[1])[0]
```
Comment (AetherUnbound) [inline tests/test_cpmorphology.py]: What's the reason for this change?

**Comment by AetherUnbound** [comment]
My greatest concern is backwards compatibility, since changing all of CellProfiler to Python 3 will be a large undertaking. It looks like all the tests (but bilateral, as you described) is passing! @0x00b1 may have to look into that a bit more. This LGTM, but I want to defer to Allen for merges on this repo.

The only other thing I would say is `2to3` adds `list` castings around generators that don't necessarily need to be converted to lists. They're sprinkled throughout your repo so I didn't want to comment on every one, but many of those are likely unnecessary. 

Thanks for your work! This is an awesome set of changes 🎉

File: centrosome/neighmovetrack.py (hunk 3 above)
Comment (jni) [reply centrosome/neighmovetrack.py]: HOLY MOLY!

I had no idea that:
- modernize could make changes that actually break the code
- `def func((a, b), c)` was valid Python
- that this was changed in Python 3

My general approach to Python 3 compatibility is:
- run `python-modernize -w -n .` on a codebase
- fix all the failing tests

This means that many redundant `list()` wrappers will remain and that anything broken that is not covered by a test will remain broken. That's what happened here and in many of the redundant wrappers. (Sometimes I remove these but only when they are completely obvious and I run into them during my test repairs.)

I'll update this *and* add a test. ;)

File: tests/test_cpmorphology.py (hunk 5 above)
Comment (jni) [reply tests/test_cpmorphology.py]: @AetherUnbound `calculate_extents` returns an array. I'm not sure whether there was a change in the behaviour of NumPy or unittest (Py3) since this test was written, I presume it's NumPy, but currently you can't use `unittest.TestCase.assertAlmostEqual` for *approximate* equality of an array of shape (1,) to a number:

```python
In [14]: c = unittest.TestCase()

In [15]: result = np.pi / 4 + np.random.random(1) / 400 - (1/800)

In [16]: result
Out[16]: array([ 0.7859997])

In [17]: np.pi / 4
Out[17]: 0.7853981633974483

In [18]: c.assertAlmostEqual(result, np.pi / 4, 2)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-cec72c7712b5> in <module>()
----> 1 c.assertAlmostEqual(result, np.pi / 4, 2)

~/conda/envs/36/lib/python3.6/unittest/case.py in assertAlmostEqual(self, first, second, places, msg, delta)
    860                 places = 7
    861
--> 862             if round(abs(second-first), places) == 0:
    863                 return
    864

TypeError: type numpy.ndarray doesn't define __round__ method
```

Grabbing that element works:

```python
In [19]: c.assertAlmostEqual(result[0], np.pi / 4, 2)
```

(no output, as expected)

Amusingly/confusingly, `assertAlmostEqual` checks for exact equality first, which is why other similar test cases pass:

```python
In [20]: c.assertAlmostEqual(np.array([1]), 1, 2)
```

I think NumPy might be considering removing the implicit conversion of `np.array([True])` to `True`, so it would be good to fix *all* similar tests, but I consider that beyond the scope of this PR. ;)

File: centrosome/neighmovetrack.py (hunk 3 above)
Comment (jni) [reply centrosome/neighmovetrack.py]: Oh I just saw that actually the fix was not breaking, it did the right thing, it just didn't name the variables nicely. Whew! That's a load off my mind. =) I might suggest that concatenating the variable names in the tuple would be a better strategy. =)

**Comment by jni** [comment]
Ok, this should be *really* ready now except for the bilateral filter test. @0x00b1 can you investigate that one?

@AetherUnbound et al, yes, the output of modernize on the CP repo is eye-watering. 😂 😭 😂 Nevertheless, we want to expand CellProfiler use at my new group at Monash Micro Imaging so I anticipate a lot of work on the source. My OCD prevents me from working on a Python 2-only codebase without at least *attempting* an update, and anyway I would expect that CP will continue to be used well past 2020 so this work needs to happen at some point. So, I'm gonna give it a go, once this is merged. Wish me luck! 😂

File: centrosome/neighmovetrack.py [hunk 6]
```diff
@@ -1,13 +1,38 @@
...
+from . import lapjv
+import six
+from six.moves import range
+from six.moves import zip
 
 invalid_match = 1000000  # limiting the choices of the algorithms
 
 
-def euclidean_dist((x1, y1), (x2, y2)):
+def euclidean_dist(point1, point2):
```
Comment (0x00b1) [inline centrosome/neighmovetrack.py]: Nice! 👍

File: setup.py [hunk 7]
```diff
@@ -140,5 +137,5 @@ def run_tests(self):
         "pytest",
     ],
     url="https://github.com/CellProfiler/centrosome",
-    version="1.0.9",
+    version="1.1.0-dev",
```
Comment (0x00b1) [inline setup.py]: We should go ahead and use `1.1.0` (or even `2.0.0`).

File: tests/test_cpmorphology.py (hunk 5 above)
Comment (0x00b1) [reply tests/test_cpmorphology.py]: Makes sense! Thanks for the explanation. 👍

**Comment by 0x00b1** [comment]
> The only other thing I would say is 2to3 adds list castings around generators that don't necessarily need to be converted to lists. They're sprinkled throughout your repo so I didn't want to comment on every one, but many of those are likely unnecessary.

Yeah, I was curious about that too. Nevertheless, it doesn’t really bother me. Do you have a preference @AetherUnbound?

**Comment by 0x00b1** [comment]
> Ok, this should be really ready now except for the bilateral filter test. @0x00b1 can you investigate that one?

Yep! No problem! @bethac07 and I are looking into this now. I really appreciate the PR!

**Comment by 0x00b1** [comment]
@jni @AetherUnbound: @bethac07 and I deprecated `bilateral_filter` for `skimage.restoration.denoise_bilateral`:

https://github.com/jni/centrosome/pull/1

**Comment by jni** [comment]
@0x00b1 

> Yeah, I was curious about that too. Nevertheless, it doesn’t really bother me. Do you have a preference @AetherUnbound?

I think I removed all of the unnecessary casts in 4dd28fc anyway...

File: setup.py (hunk 7 above)
Comment (jni) [reply setup.py]: @0x00b1 at scikit-image we prefer to use -dev except for release branches, so if someone says "v0.13.1" we know they are not on some weird between-releases commit. But, no problem, bumped to 2.0.0. =)

**Comment by jni** [comment]
Hmm, bilateral filter is failing in Py2.7 anyway... Maybe it's a NumPy change more than a Python version change...?

**Comment by jni** [comment]
@AetherUnbound @0x00b1 care to take another look at merging this?

**Comment by jni** [comment]
ping @AetherUnbound @0x00b1

**Comment by jni** [comment]
ping some more! =P Just fyi I'm back from SciPyConf and would like to start hacking at CellProfiler sometime next week.

**Comment by jni** [comment]
:tada:
